Add `unique_indices` arg and update nondeterministic alerts for `scatter` and `scatter_add`

kurtamohler opened the pull request:
Fixes #70583


cc @mruberry @kurtamohler

kurtamohler committed:
Add nondeterministic alert to `scatter`

kurtamohler commented:
I suppose `scatter_add` should also always alert nondeterministic as well, right? At the moment, it only alerts on CUDA

Unless I'm mistaken, I think we should also make `scatter_add` always alert nondeterministic. At the moment, it's only alerting on CUDA if the `input` has more than one dimension, because (if I remember correctly) `gpuAtomicAdd` is only used in that case. But regardless of that, `scatter_add` should also have the duplicate indices issue. I can fix the `scatter_add` alerts in this PR too if that makes sense.

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -143,11 +143,17 @@ void scatter_meta_impl(
 (const Tensor& self, int64_t dim, const Tensor& index, const Tensor& src) {
+  // See note [Writing Nondeterministic Operations]
+  // Nondeterministic when index contains duplicate entries
+  at::globalContext().alertNotDeterministic("scatter_src_out");
```
In the future should we think about letting the users know information like, "This operation is deterministic if the input indices are unique, and this can be validated by..."? 

Or is it preferable to just implement the best validation of unique input indices as the deterministic version of scatter?

kurtamohler replied:
Yeah it might be a good idea to let users know the conditions that make an operation nondeterministic.

> implement the best validation of unique input indices as the deterministic version of scatter

Do you mean something like this?

```
if (using deterministic algorithms) and (index contains duplicate entries):
    throw error
else:
    run operation
```

mruberry commented:
Test failure appears to be https://github.com/pytorch/pytorch/issues/71095

mruberry reviewed (approved):
Cool! Thanks @kurtamohler

mruberry commented:
Not super surprisingly the popularity of this flag means that this is going to break some internal tests where people are validating networks continue to run, but those networks are using this flag + scatter.

I'll reach out to the affected test owners and ask them to update their tests. An alternative would be to implement a deterministic version of CUDA scatter with this PR, although I don't have a great sense for how challenging that would be.

mruberry commented:
> 

So internal uses of deterministic scatter are open-minded about revising their tests, but after chatting with them and @ngimel I'm worried about the impact of this change.

Instead of just marking scatter nondeterministic should we adopt the `unique_indices` kwarg from JAX? If a user promises that indices are unique then the operation is deterministic, right? I think that might make this change a lot easier for deterministic scatter users to adapt to, instead of having to act as if scatter might be nondeterministic (but they know it isn't based on their inputs).

ngimel commented:
Concretely, in the tests that are currently failing they are getting indices from `topk` or `sort` operation, so indices are guaranteed to be unique, and use those indices in `scatter`. There could be other cases where user has a knowledge of indices properties. So yes, adding an argument to `scatter` and `scatter_reduce` will be helpful.

kurtamohler commented:
Just trying to make sure I understand the idea. When this `unique_indices` argument is True, then `scatter` would not throw a nondeterministic error when `use_deterministic_algorithms(True)`? If a user actually does provide duplicate indices, even though they're not supposed to, then `scatter` is free to act nondeterministically without error.

It seems questionable to have this hidden nondeterministic behavior. To me, it feels like it would conflict with the purpose of `use_deterministic_algorithms`, which is to help people either automatically avoid nondeterminism or to point out exactly where nondeterminism is happening, for every operation that we know to be nondeterministic.

Also, instead of `unique_indices`, a user could just turn off the determinism flag temporarily:

```python
torch.use_deterministic_algorithms(False)
torch.scatter(...)
torch.use_deterministic_algorithms(True)
```

I believe this achieves the same thing, and it seems better since there is no conceptual conflict. Could the internal tests just do this as well?

Side note: If I understand the JAX documentation correctly, it seems like their `unique_indices` argument is purely performance-related. So if the user provides duplicate indices, it doesn't seem like anything incorrect happens, it just might run more slowly than normal. It may or may not affect determinism in JAX, but it doesn't matter for them since they don't offer a determinism flag (at least, I don't see one in the docs).

mruberry commented:
> Just trying to make sure I understand the idea. When this `unique_indices` argument is True, then `scatter` would not throw a nondeterministic error when `use_deterministic_algorithms(True)`? If a user actually does provide duplicate indices, even though they're not supposed to, then `scatter` is free to act nondeterministically without error.

Correct. We would need to update the documentation to explain that if `unique_indices=True` but indices are not unique then the operation's behavior is undefined, I think.

> Even if the docs forbid the user from calling `scatter` this way, they still might do it. It seems questionable to allow this hidden nondeterministic behavior to happen. To me, it feels like it would conflict with the purpose of `use_deterministic_algorithms`, which is to either automatically avoid nondeterminism or to point out exactly where nondeterminism is happening, for every operation that we know to be nondeterministic.

This is a completely reasonable way to look at the deterministic flag (as avoiding determinism at any cost), although I think the user making a promise about the input is also valid to respect? Our nondeterministic algorithms are not nondeterministic when given overlapped tensors, either, but the documentation warns the user that these tensors are a problem and not to pass them to our ops.

> Also, instead of `unique_indices`, a user could just turn off the determinism flag temporarily:
> 
> ```python
> torch.use_deterministic_algorithms(False)
> torch.scatter(...)
> torch.use_deterministic_algorithms(True)
> ```
> 
> I believe this achieves the same thing, and it seems better since there is no conceptual conflict. Could the internal tests just do this as well?

Absolutely and test owners so far seem OK with this workaround. @ngimel and I just thought it was odd that in the cases we saw the calls were actually provably deterministic, but we were still requiring them to mark the section as (potentially) nondeterministic. As you mention above that kind a felt like these areas were highlighting nondeterministic behavior, but that's not what's happening.

> Side note: If I understand the JAX documentation correctly, it seems like their `unique_indices` argument is purely performance-related. So if the user provides duplicate indices, it doesn't seem like anything incorrect happens, it just might run more slowly than normal (though I haven't actually tried it, so I don't know this for sure). It may or may not affect determinism in JAX, but it doesn't matter for them since they don't offer a determinism flag (at least, I don't see one in the docs).

I don't see how JAX's scatter could work correctly all the time if `unique_indices` is set to True but there are actually duplicate indices. It'd be interesting to look at some examples, however.

kurtamohler commented:
> that kind a felt like these areas were highlighting nondeterministic behavior, but that's not what's happening.

That's a very good point. Alright, I agree with you both now, thanks for engaging my skepticism

kurtamohler committed:
Add `unique_indices` arg to `scatter` and `scatter_add`

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler commented:
I've added the `unique_indices` arg to `scatter` and `scatter_add`, and updated them to avoid throwing a nondeterministic error due to possible duplicate indices when `unique_indices=True`. I also added a new nondeterministic alert to `scatter(..., reduction='add')` due to its usage of `gpuAtomicAdd`.

Something else needs to be addressed: `gather` calls `scatter_add` internally. At the moment, `unique_indices` is always False for that call, and `gather` is always throwing the nondeterministic error now. So I think I should probably add `unique_indices` to `gather` as well

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1300,18 +1300,18 @@
   result: auto_element_wise
 
-- name: scatter.src(Tensor self, int dim, Tensor index, Tensor src) -> Tensor
+- name: scatter.src(Tensor self, int dim, Tensor index, Tensor src, *, bool unique_indices=False) -> Tensor
```
Since the same indices are re-used during the backward, should this flag be propagated?

kurtamohler replied:
You're right, good catch

kurtamohler committed:
Add `unique_indices` to `gather` and `scatter` backward formulas

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler committed:
Update documentation and autograd nondeterministic error tests

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler committed:
Fix JIT issue

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler committed:
Attempt to fix FC error with gather_backward

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler committed:
Attempt to fix ONNX errors

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -221,19 +221,19 @@ def __interpolate(g, input, size, scale_factor, mode, align_corners, recompute_s
-def gather(g, self, dim, index, sparse_grad=False):
+def gather(g, self, dim, index, sparse_grad=False, unique_indices=False):
     if sym_help._maybe_get_const(sparse_grad, "i"):
         return _unimplemented("gather", "sparse_grad == True")
```
I'm not sure whether or not I should add `_unimplemented()` calls for `unique_indices == True`. I'm not very familiar with the ONNX stuff

kurtamohler replied:
I decided to just add the `_unimplemented()` calls. I can remove them if needed

kurtamohler committed:
Mark `unique_indices == True` unimplimented for ONNX

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler committed:
Another attempt to fix ONNX error

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler commented:
Hey @JackCaoG, this PR is adding a kwarg to some functions, so it's causing these XLA errors:

<details>
<summary>Click to expand/collapse</summary>

```
2022-02-09T21:02:57.6253654Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:1487:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::gather(const at::Tensor&, int64_t, const at::Tensor&, bool)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6254506Z  at::Tensor XLANativeFunctions::gather(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6254940Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6255348Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6256147Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:179:19: error: candidate is: static at::Tensor torch_xla::XLANativeFunctions::gather(const at::Tensor&, int64_t, const at::Tensor&, bool, bool)
2022-02-09T21:02:57.6256943Z  static at::Tensor gather(const at::Tensor & self, int64_t dim, const at::Tensor & index, bool sparse_grad, bool unique_indices);
2022-02-09T21:02:57.6257397Z                    ^~~~~~
2022-02-09T21:02:57.6258501Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp: In function ‘at::Tensor torch_xla::scatter_reduce_helper(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::optional<c10::basic_string_view<char> >)’:
2022-02-09T21:02:57.6260572Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2852:73: error: no matching function for call to ‘at::native::_call_fallback_fn<torch_xla::xla_cpu_fallback, at::_ops::scatter_reduce, at::Tensor(const at::Tensor&, long int, const at::Tensor&, const at::Tensor&, c10::basic_string_view<char>, bool)>::call(const at::Tensor&, int64_t&, const at::Tensor&, const at::Tensor&, c10::basic_string_view<char>&)’
2022-02-09T21:02:57.6261628Z                                                              src, *reduce);
2022-02-09T21:02:57.6261974Z                                                                          ^
2022-02-09T21:02:57.6262467Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:4:0:
2022-02-09T21:02:57.6264540Z /opt/conda/lib/python3.7/site-packages/torch/include/ATen/native/CPUFallback.h:28:23: note: candidate: static ReturnType at::native::_call_fallback_fn<fallback_fn, Op, ReturnType(ParameterTypes ...)>::call(ParameterTypes ...) [with void (* fallback_fn)(const c10::OperatorHandle&, c10::Stack*) = torch_xla::xla_cpu_fallback; Op = at::_ops::scatter_reduce; ReturnType = at::Tensor; ParameterTypes = {const at::Tensor&, long int, const at::Tensor&, const at::Tensor&, c10::basic_string_view<char>, bool}]
2022-02-09T21:02:57.6265844Z      static ReturnType call(ParameterTypes... args) {
2022-02-09T21:02:57.6266234Z                        ^~~~
2022-02-09T21:02:57.6267007Z /opt/conda/lib/python3.7/site-packages/torch/include/ATen/native/CPUFallback.h:28:23: note:   candidate expects 6 arguments, 5 provided
2022-02-09T21:02:57.6268512Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp: In function ‘at::Tensor torch_xla::scatter_reduce_helper(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::optional<c10::basic_string_view<char> >)’:
2022-02-09T21:02:57.6270491Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2873:74: error: no matching function for call to ‘at::native::_call_fallback_fn<torch_xla::xla_cpu_fallback, at::_ops::scatter_value_reduce, at::Tensor(const at::Tensor&, long int, const at::Tensor&, const c10::Scalar&, c10::basic_string_view<char>, bool)>::call(const at::Tensor&, int64_t&, const at::Tensor&, const c10::Scalar&, c10::basic_string_view<char>&)’
2022-02-09T21:02:57.6271509Z                                                                    *reduce);
2022-02-09T21:02:57.6271892Z                                                                           ^
2022-02-09T21:02:57.6272364Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:4:0:
2022-02-09T21:02:57.6274510Z /opt/conda/lib/python3.7/site-packages/torch/include/ATen/native/CPUFallback.h:28:23: note: candidate: static ReturnType at::native::_call_fallback_fn<fallback_fn, Op, ReturnType(ParameterTypes ...)>::call(ParameterTypes ...) [with void (* fallback_fn)(const c10::OperatorHandle&, c10::Stack*) = torch_xla::xla_cpu_fallback; Op = at::_ops::scatter_value_reduce; ReturnType = at::Tensor; ParameterTypes = {const at::Tensor&, long int, const at::Tensor&, const c10::Scalar&, c10::basic_string_view<char>, bool}]
2022-02-09T21:02:57.6275773Z      static ReturnType call(ParameterTypes... args) {
2022-02-09T21:02:57.6276148Z                        ^~~~
2022-02-09T21:02:57.6276904Z /opt/conda/lib/python3.7/site-packages/torch/include/ATen/native/CPUFallback.h:28:23: note:   candidate expects 6 arguments, 5 provided
2022-02-09T21:02:57.6277602Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp: At global scope:
2022-02-09T21:02:57.6279037Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2877:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6279959Z  at::Tensor XLANativeFunctions::scatter(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6280383Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6280969Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6281856Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:283:19: error: candidates are: static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::string_view, bool)
2022-02-09T21:02:57.6282933Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6283459Z                    ^~~~~~~
2022-02-09T21:02:57.6284167Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:282:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, bool)
2022-02-09T21:02:57.6285045Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, bool unique_indices);
2022-02-09T21:02:57.6285562Z                    ^~~~~~~
2022-02-09T21:02:57.6286304Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:281:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::string_view, bool)
2022-02-09T21:02:57.6287259Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6287811Z                    ^~~~~~~
2022-02-09T21:02:57.6288510Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:280:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, bool)
2022-02-09T21:02:57.6289378Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, bool unique_indices);
2022-02-09T21:02:57.6289896Z                    ^~~~~~~
2022-02-09T21:02:57.6291220Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2884:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6292113Z  at::Tensor XLANativeFunctions::scatter(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6292529Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6292951Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6294013Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:283:19: error: candidates are: static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::string_view, bool)
2022-02-09T21:02:57.6295170Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6295701Z                    ^~~~~~~
2022-02-09T21:02:57.6296423Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:282:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, bool)
2022-02-09T21:02:57.6297385Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, bool unique_indices);
2022-02-09T21:02:57.6297890Z                    ^~~~~~~
2022-02-09T21:02:57.6298647Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:281:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::string_view, bool)
2022-02-09T21:02:57.6299628Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6300154Z                    ^~~~~~~
2022-02-09T21:02:57.6300841Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:280:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, bool)
2022-02-09T21:02:57.6301771Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, bool unique_indices);
2022-02-09T21:02:57.6302271Z                    ^~~~~~~
2022-02-09T21:02:57.6303641Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2891:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::string_view)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6304569Z  at::Tensor XLANativeFunctions::scatter(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6304983Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6305453Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6306352Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:283:19: error: candidates are: static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::string_view, bool)
2022-02-09T21:02:57.6307359Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6307904Z                    ^~~~~~~
2022-02-09T21:02:57.6308591Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:282:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, bool)
2022-02-09T21:02:57.6309513Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, bool unique_indices);
2022-02-09T21:02:57.6310030Z                    ^~~~~~~
2022-02-09T21:02:57.6310728Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:281:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::string_view, bool)
2022-02-09T21:02:57.6311666Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6312168Z                    ^~~~~~~
2022-02-09T21:02:57.6312857Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:280:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, bool)
2022-02-09T21:02:57.6313857Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, bool unique_indices);
2022-02-09T21:02:57.6314385Z                    ^~~~~~~
2022-02-09T21:02:57.6315732Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2899:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::string_view)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6316752Z  at::Tensor XLANativeFunctions::scatter(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6317188Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6317634Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6318523Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:283:19: error: candidates are: static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, c10::string_view, bool)
2022-02-09T21:02:57.6319469Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6319987Z                    ^~~~~~~
2022-02-09T21:02:57.6320687Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:282:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, bool)
2022-02-09T21:02:57.6321571Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, bool unique_indices);
2022-02-09T21:02:57.6322052Z                    ^~~~~~~
2022-02-09T21:02:57.6322768Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:281:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, c10::string_view, bool)
2022-02-09T21:02:57.6323728Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, c10::string_view reduce, bool unique_indices);
2022-02-09T21:02:57.6324246Z                    ^~~~~~~
2022-02-09T21:02:57.6324909Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:280:19: error:                 static at::Tensor torch_xla::XLANativeFunctions::scatter(const at::Tensor&, int64_t, const at::Tensor&, const c10::Scalar&, bool)
2022-02-09T21:02:57.6325778Z  static at::Tensor scatter(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Scalar & value, bool unique_indices);
2022-02-09T21:02:57.6326278Z                    ^~~~~~~
2022-02-09T21:02:57.6327609Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:2907:12: error: prototype for ‘at::Tensor torch_xla::XLANativeFunctions::scatter_add(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&)’ does not match any in class ‘torch_xla::XLANativeFunctions’
2022-02-09T21:02:57.6328573Z  at::Tensor XLANativeFunctions::scatter_add(const at::Tensor& self, int64_t dim,
2022-02-09T21:02:57.6329010Z             ^~~~~~~~~~~~~~~~~~
2022-02-09T21:02:57.6329458Z In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:13:0:
2022-02-09T21:02:57.6330330Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/XLANativeFunctions.h:284:19: error: candidate is: static at::Tensor torch_xla::XLANativeFunctions::scatter_add(const at::Tensor&, int64_t, const at::Tensor&, const at::Tensor&, bool)
2022-02-09T21:02:57.6331218Z  static at::Tensor scatter_add(const at::Tensor & self, int64_t dim, const at::Tensor & index, const at::Tensor & src, bool unique_indices);
2022-02-09T21:02:57.6331721Z                    ^~~~~~~~~~~
```

</details>

wonjoolee95 commented:
PR for this fix (https://github.com/pytorch/xla/pull/3358) currently under review on PT/XLA's side.

wonjoolee95 commented:
Quick update: the PR has been approved, now we're waiting on our CircleCI tests to pass. When they all pass, I'll message here again to let you know. Thanks!

ngimel commented on torch/_tensor_docs.py:
```diff
@@ -3291,8 +3292,15 @@ def callable(a, b) -> number
+Keyword arguments:
     reduce (str, optional): reduction operation to apply, can be either
         ``'add'`` or ``'multiply'``.
+    unique_indices (bool, optional): whether the indices to be updated in
```
can you add that it's users responsibility to correctly set `unique_indices` and it's not checked if indices are really unique?

kurtamohler replied:
done

wonjoolee95 commented:
@kurtamohler, PT/XLA fix https://github.com/pytorch/xla/pull/3358 is green. From PT/XLA's side, we should be good to merge this change.

kurtamohler committed:
Clarify in docs that indices are not check for overlaps

kurtamohler committed:
Merge branch 'master' into scatter-nondeterministic-alert-70583

kurtamohler commented:
@mruberry any updates on importing this?

kit1980 commented:
/easycla
